File: dotnet4-x/workspaces/warehouse/src/Core/Warehouse.cs
using System;
using System.Collections.Generic;
using System.Threading;

namespace App
{
    class Warehouse
    {
        private Dictionary<int, Item> items;
        private object lockObject;

        public Warehouse()
        {
            this.items = new Dictionary<int, Item>();
            this.lockObject = new object();
        }

        public void AddItem(Item item)
        {
            lock (this.lockObject)
            {
                this.items[item.Id] = item;
            }
        }

        public bool Reserve(int id, int count)
        {
            lock (this.lockObject)
            {
                if (this.items.ContainsKey(id))
                {
                    Item item = this.items[id];
                    if (item.Quantity >= count)
                    {
                        item.Quantity = item.Quantity - count;
                        return true;
                    }
                }
                return false;
            }
        }

        public int QuantityOf(int id)
        {
            lock (this.lockObject)
            {
                if (this.items.ContainsKey(id))
                {
                    return this.items[id].Quantity;
                }
                return 0;
            }
        }
    }
}

File: Windows/cpp/workspaces/warehouse/src/core/Warehouse.cpp
#include "Warehouse.h"
#include <algorithm>

namespace core {
    Warehouse::Warehouse() = default;

    Warehouse::~Warehouse() = default;

    void Warehouse::AddItem(const Item& item) {
        std::lock_guard<std::mutex> lock(mutex_);
        items_.insert_or_assign(item.getId(), item);
    }

    bool Warehouse::Reserve(int id, int count) {
        std::lock_guard<std::mutex> lock(mutex_);
        auto it = items_.find(id);
        if (it == items_.end() || it->second.getQuantity() < count) {
            return false;
        }
        it->second.setQuantity(it->second.getQuantity() - count);
        return true;
    }

    int Warehouse::QuantityOf(int id) const {
        std::lock_guard<std::mutex> lock(mutex_);
        auto it = items_.find(id);
        if (it == items_.end()) {
            return 0;
        }
        return it->second.getQuantity();
    }
} // namespace core
